DIPs/DIP1009.md

```diff
@@ -45,197 +39,323 @@ body
+Contracts can use either the statement-based or the expression-based syntax. The `do` keyword continues to be required as a separator between a statement-based contract and the function body.
 
 Example:
 ```d
 int fun(ref int a, int b)
 in(a > 0)
-in {
-      assert(b >= 0, "b cannot be negative!");
+in
+{
+    assert(b >= 0, "b cannot be negative!");
 }
-do {
-   // function body
+do
+{
+    // function body
 }
 ```
 
-In the case of body-less virtual interface functions, a semicolon is required to terminate the function declaration. This differs from the old contract syntax, which does not require a terminating semicolon for body-less functions.
+For virtual interface functions without a specified function body, a semicolon is required to terminate the function declaration after an expression-based contract specification. (No semicolon is required to terminate such a function declaration after a statement-based contract specification, which is consistent with current behaviour.)
 
 Example:
 ```d
-interface I {
-   int fun(int a) in(a > 0); 
+interface I
+{
+    int fun(int a) in(a > 0);
 }
 ```
 
-This DIP also extends the proposed syntax to `invariant` contracts. Like the proposed function contract syntax, the proposed `invariant` contract syntax is designed to mimic a call to `assert`. These modified invariant contracts must be terminated with a semicolon.
+This DIP also proposes similar syntax for `invariant` contracts. Like the proposed function contract syntax, the new `invariant` contract syntax is designed to mimic a call to `assert`.
+Expression-based `invariant` contracts must be terminated with a semicolon.
 
 Example:
 ```d
 class C {
-   int data;
-   invariant(data != 0, "cannot be 0");
+    int data;
+    invariant(data != 0, "data cannot be 0");
 }
 ```
 
-The proposed contract syntax will be lowered to the old syntax. For example, the above contracts written with the proposed syntax would be lowered as follows:
+The semantics of the new expression-based contracts is specified by lowering to contracts using the existing statement-based syntax.
+Expression-based contracts are lowered to statement-based contracts containing a single `assert` statement with the arguments of the contract.
+
+Example:
 ```d
 int fun(ref int a, int b)
-in {
-   assert(a > 0);
-   assert(b >= 0, "b cannot be negative!");
+    in(a > 0)
+    in(b >= 0, "b cannot be negative!")
+    out(r; r > 0, "return must be positive")
+    out(; a != 0)
+{
+    // function body
 }
-out(r) {
-   assert(r > 0, "return must be positive");
-   assert(a != 0);
+
+class C {
+    int data;
+    invariant(data != 0, "data cannot be 0");
+}
+```
+Turns into:
+```d
+int fun(ref int a, int b)
+in
+{
+    assert(a > 0);
+}
+in
+{
+    assert(b >= 0, "b cannot be negative!");
 }
-do {
-   // function body
+out(r)
+{
+    assert(r > 0, "return must be positive");
+}
+out
+{
+    assert(a != 0);
+}
+do
+{
+    // function body
 }
 
 interface I {
```

Comment:
I don't know if it matters at this point, but this interface is not in the above untranslated block.

Reply:
Thanks! https://github.com/dlang/DIPs/pull/116